src/wallet/sendTransaction.ts

```diff
@@ -0,0 +1,353 @@
+/**
+ * Copyright (c) Hathor Labs and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { EventEmitter } from 'events';
+import walletApi from './api/walletApi';
+import MineTransaction from './mineTransaction';
+import HathorWalletServiceWallet from './wallet';
+import Transaction from '../models/transaction';
+import Output from '../models/output';
+import Input from '../models/input';
+import Address from '../models/address';
+import { HATHOR_TOKEN_CONFIG } from '../constants';
+import { shuffle } from 'lodash';
+import { SendTxError, UtxoError, WalletError, WalletRequestError } from '../errors';
+import { OutputRequestObj, InputRequestObj } from './types';
+
+type optionsType = {
+  outputs?: OutputRequestObj[],
+  inputs?: InputRequestObj[],
+  changeAddress?: string | null,
+  transaction?: Transaction | null,
+};
+
+class SendTransactionWalletService extends EventEmitter {
+  // Wallet that is sending the transaction
+  private wallet: HathorWalletServiceWallet;
+  // Outputs to prepare the transaction
+  private outputs: OutputRequestObj[];
+  // Optional inputs to prepare the transaction
+  private inputs: InputRequestObj[];
+  // Optional change address to prepare the transaction
+  private changeAddress: string | null;
+  // Transaction object to be used after it's already prepared
+  private transaction: Transaction | null;
+  // MineTransaction object
+  private mineTransaction: MineTransaction | null;
+
+  constructor(wallet: HathorWalletServiceWallet, options: optionsType = {}) {
+    super();
+
+    const newOptions: optionsType = Object.assign({
+      outputs: [],
+      inputs: [],
+      changeAddress: null,
+      transaction: null,
+    }, options);
+
+    this.wallet = wallet;
+    this.outputs = newOptions.outputs!;
+    this.inputs = newOptions.inputs!;
+    this.changeAddress = newOptions.changeAddress!;
+    this.transaction = newOptions.transaction!;
+    this.mineTransaction = null;
+  }
+
+  /**
+   * Prepare transaction data to send
+   * Get utxos from wallet service, creates change outpus and returns a Transaction object
+   *
+   * @memberof SendTransactionWalletService
+   * @inner
+   */
+  async prepareTx(): Promise<{ transaction: Transaction, utxosAddressPath: string[] }> {
+    if (this.outputs.length === 0) {
+      throw new WalletError('Can\'t prepare transactions with no outputs.');
+    }
+    this.emit('prepare-tx-start');
+    // We get the full outputs amount for each token
+    // This is useful for (i) getting the utxos for each one
+    // in case it's not sent and (ii) create the token array of the tx
+    const amountOutputMap = {};
+    for (const output of this.outputs) {
+      if (output.token in amountOutputMap) {
+        amountOutputMap[output.token] += output.value;
+      } else {
+        amountOutputMap[output.token] = output.value;
+      }
+    }
+
```

Comment:
**Comment unrelated to line number, but related to the current method**

I feel that this method is big, we should add more comments explaining each step starting from this line

Reply:
I agree. Instead of comments, I feel we should split it in several methods that can be independently tested.

Reply:
I added more comments and split the method into some smaller ones. Can you check please?
https://github.com/HathorNetwork/hathor-wallet-lib/pull/221/commits/d7e7446efd4b3b7ab2f1e026a12f8f9427baadf0

Reply:
Looking great to me :+1:

Reply:
Well done. Thanks!